File: oracle_connector/face_measurement.py
"""Validate bounded browser measurements, bound to the exact uploaded JPEG."""
import math


def validate(value, width, height, digest):
    if value is None:
        return None
    if not isinstance(value,dict) or value.get('revision')!=1 or value.get('width')!=width or value.get('height')!=height or value.get('imageSha256')!=digest:
        raise ValueError('Pomiary twarzy nie pasuja do zdjecia. Dodaj je ponownie.')
    points=value.get('points')
    if not isinstance(points,list) or len(points)!=478:
        raise ValueError('Niepelne pomiary twarzy.')
    for point in points:
        if not isinstance(point,list) or len(point)!=3 or any(type(x) not in (int,float) or not math.isfinite(x) or abs(x)>2 for x in point):
            raise ValueError('Nieprawidlowe wspolrzedne twarzy.')
    return {key:value[key] for key in ('revision','width','height','imageSha256','points')}

File: src/blender/faceMeasurement.ts
export type FaceMeasurement = { revision: 1; width: number; height: number; imageSha256: string; points: number[][] }

export function validateFaceMeasurement(value: unknown, width: number, height: number, sha256: string): FaceMeasurement | undefined {
  if (value === undefined) return undefined
  const v = value as FaceMeasurement
  if (!v || v.revision !== 1 || v.width !== width || v.height !== height || v.imageSha256 !== sha256 ||
      !Array.isArray(v.points) || v.points.length !== 478 || v.points.some(p => !Array.isArray(p) || p.length !== 3 || p.some(x => typeof x !== 'number' || !Number.isFinite(x) || Math.abs(x) > 2))) {
    throw new Error('Pomiary twarzy nie pasują do zdjęcia. Dodaj je ponownie.')
  }
  return { revision: 1, width, height, imageSha256: sha256, points: v.points }
}
